dynamic/internal/shim/protov5/translate/tfplugin5.go

```diff
@@ -0,0 +1,674 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translate
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/opentofu/opentofu/internal/tfplugin5"
```

Comment:
Ah two models for the same data, how sad. 

Are you sure you need these translations?

E.g.  tfprotov5.GetMetadataRequest -> tfplugin5.GetMetadata_Request

These are all protobuf-level models I think. So you could turn around through proto:

tfprotov5.GetMetadataRequest -> proto-bytes -> tfplugin5.GetMetadata_Request

Reply:
They both represent the same thing, but they aren't both protobuf models. tfprotov5 is just slightly higher level then tfplugin5.

For example, `tfplugin5.ServerCapabilities` vs `*tfprotov5.ServerCapabilities`:

```go
// ServerCapabilities allows providers to communicate extra information
// regarding supported protocol features. This is used to indicate
// availability of certain forward-compatible changes which may be optional
// in a major protocol version, but cannot be tested for directly.
type ServerCapabilities struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// The plan_destroy capability signals that a provider expects a call
	// to PlanResourceChange when a resource is going to be destroyed.
	PlanDestroy bool `protobuf:"varint,1,opt,name=plan_destroy,json=planDestroy,proto3" json:"plan_destroy,omitempty"`
	// The get_provider_schema_optional capability indicates that this
	// provider does not require calling GetProviderSchema to operate
	// normally, and the caller can used a cached copy of the provider's
	// schema.
	GetProviderSchemaOptional bool `protobuf:"varint,2,opt,name=get_provider_schema_optional,json=getProviderSchemaOptional,proto3" json:"get_provider_schema_optional,omitempty"`
	// The move_resource_state capability signals that a provider supports the
	// MoveResourceState RPC.
	MoveResourceState bool `protobuf:"varint,3,opt,name=move_resource_state,json=moveResourceState,proto3" json:"move_resource_state,omitempty"`
}
```

```go
// ServerCapabilities allows providers to communicate optionally supported
// protocol features, such as forward-compatible Terraform behavior changes.
//
// This information is used in GetProviderSchemaResponse as capabilities are
// static features which must be known upfront in the provider server.
type ServerCapabilities struct {
	// GetProviderSchemaOptional signals that this provider does not require
	// having the GetProviderSchema RPC called first to operate normally. This
	// means the caller can use a cached copy of the provider's schema instead.
	GetProviderSchemaOptional bool

	// MoveResourceState signals that a provider supports the MoveResourceState
	// RPC.
	MoveResourceState bool

	// PlanDestroy signals that a provider expects a call to
	// PlanResourceChange when a resource is going to be destroyed. This is
	// opt-in to prevent unexpected errors or panics since the
	// ProposedNewState in PlanResourceChangeRequest will be a null value.
	PlanDestroy bool
}
```

I thought of trying to automate the mapping with reflect, but there are a couple of places where you need genuine type conversions:

https://github.com/pulumi/pulumi-terraform-bridge/blob/114bc771e674f48f61b071bffec0456cc7ff214f/dynamic/internal/shim/protov5/translate/tfplugin5.go#L190-L202